src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns_Constant.cs

```diff
@@ -0,0 +1,159 @@
+            // PROTOTYPE(patterns): an integral type, but we currently treat all inputs the same as `object`.
+            var value = boundConstant.ConstantValue;
+            if (!value.IsIntegral)
+            {
+                // We use object.Equals for char, string, float, double, decimal, DateTime, and null
+                return _factory.StaticCall(
+                    _factory.SpecialType(SpecialType.System_Object),
+                    // PROTOTYPE(patterns): Add a special member for `public static bool object.Equals(object, object)`
+                    "Equals",
+                    _factory.Convert(_factory.SpecialType(SpecialType.System_Object), input),
+                    _factory.Convert(_factory.SpecialType(SpecialType.System_Object), boundConstant)
+                    );
+            }
+
+            // We use one of three helper functions from PrivateImplementationDetails, depending on the value
+            // of the constant. The helper function takes the input and produces an integral value that we
+            // are to compare against the input. Then we generate code like so:
+            //     SelectedHelperMethod((object)input, out var temp) && temp == Constant
+            MethodSymbol helper = SelectIntegralPatternHelper(boundConstant);
+            if (helper == null)
+            {
+                Debug.Assert(this.EmitModule == null);
+                return BadExpression(input);
+            }
+
+            Debug.Assert(helper.Parameters[0].Type.SpecialType == SpecialType.System_Object);
+            var comparisonType = helper.Parameters[1].Type;
+            var temp = _factory.SynthesizedLocal(comparisonType);
+            var objectType = _factory.SpecialType(SpecialType.System_Object);
+            var call = _factory.Call(null, helper,
+                ImmutableArray.Create(RefKind.None, RefKind.Out),
+                ImmutableArray.Create(_factory.Convert(objectType, input), _factory.Local(temp)));
+            BinaryOperatorKind comparison;
+            switch (comparisonType.SpecialType)
+            {
+                case SpecialType.System_Int32: comparison = BinaryOperatorKind.IntEqual; break;
+                case SpecialType.System_Int64: comparison = BinaryOperatorKind.LongEqual; break;
+                case SpecialType.System_UInt64: comparison = BinaryOperatorKind.ULongEqual; break;
+                default: throw ExceptionUtilities.UnexpectedValue(comparisonType.Name);
+            }
+
+            var equalsTest =
+                _factory.Binary(comparison, _factory.SpecialType(SpecialType.System_Boolean),
+                    _factory.Convert(comparisonType, _factory.Local(temp)),
+                    _factory.Convert(comparisonType, boundConstant));
+            return _factory.Sequence(temp, _factory.LogicalAnd(call, equalsTest));
+        }
+
+        private MethodSymbol SelectIntegralPatternHelper(BoundExpression boundConstant)
```

Comment:
On second thought, if those cases don't appear anywhere else, it's not useful to materialize.